lib/connections/service.go

```diff
@@ -336,176 +338,203 @@ func (s *service) handle(ctx context.Context) error {
+	// minConnectionLoopSleep), increased exponentially until it reaches
+	// stdConnectionLoopSleep, at which time the normal sleep mechanism
+	// kicks in.
 	initialRampup := time.Second
 
-	// Calculated from actual dialers reconnectInterval
-	var sleep time.Duration
-
 	for {
 		cfg := s.cfg.RawCopy()
+		bestDialerPrio := s.bestDialerPrio(cfg)
+		isInitialRampup := initialRampup < stdConnectionLoopSleep
 
-		bestDialerPrio := 1<<31 - 1 // worse prio won't build on 32 bit
-		for _, df := range dialers {
-			if df.Valid(cfg) != nil {
-				continue
-			}
-			if prio := df.Priority(); prio < bestDialerPrio {
-				bestDialerPrio = prio
-			}
+		l.Debugln("Connection loop")
+		if isInitialRampup {
+			l.Debugln("Connection loop in initial rampup")
 		}
 
-		l.Debugln("Reconnect loop")
-
+		// Used for consistency throughout this loop run, as time passes
+		// while we try connections etc.
 		now := time.Now()
-		var seen []string
 
-		for _, deviceCfg := range cfg.Devices {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			default:
-			}
+		// Attempt to dial all devices that are unconnected or can be connection-upgraded
+		s.dialDevices(ctx, now, cfg, bestDialerPrio, nextDial, isInitialRampup)
 
-			deviceID := deviceCfg.DeviceID
-			if deviceID == s.myID {
-				continue
-			}
+		// The sleep time is until the next dial scheduled in nextDial,
+		// clamped by stdConnectionLoopSleep as we don't want to sleep too
+		// long (config changes might happen).
+		sleep := filterAndFindSleepDuration(nextDial, now)
 
-			if deviceCfg.Paused {
-				continue
-			}
+		// .. unless we are un the initial rampup time, in which case we
+		// override the sleep interval for the first few passes through the
+		// loop.
+		if isInitialRampup {
+			sleep = initialRampup
+			initialRampup *= 2
+		}
 
-			ct, connected := s.model.Connection(deviceID)
+		// ... while making sure not to loop to quickly either.
+		if sleep < minConnectionLoopSleep {
+			sleep = minConnectionLoopSleep
+		}
 
-			if connected && ct.Priority() == bestDialerPrio {
-				// Things are already as good as they can get.
-				continue
-			}
+		l.Debugln("Next connection loop in", sleep)
 
-			var addrs []string
-			for _, addr := range deviceCfg.Addresses {
-				if addr == "dynamic" {
-					if s.discoverer != nil {
-						if t, err := s.discoverer.Lookup(ctx, deviceID); err == nil {
-							addrs = append(addrs, t...)
-						}
-					}
-				} else {
-					addrs = append(addrs, addr)
-				}
-			}
+		select {
+		case <-time.After(sleep):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
 
-			addrs = util.UniqueTrimmedStrings(addrs)
+func (s *service) bestDialerPrio(cfg config.Configuration) int {
+	first := true
+	var bestDialerPrio int
+	for _, df := range dialers {
+		if df.Valid(cfg) != nil {
+			continue
+		}
+		if prio := df.Priority(); first || prio < bestDialerPrio {
+			first = false
+			bestDialerPrio = prio
+		}
+	}
+	return bestDialerPrio
```

Comment:
This now returns zero for no dialer vs highest possible int32 before. It won't change any behaviour though, as the only condition using `bestDialerPrio` is a `==`, which will never match either 0 or highest int32. Still no dialer resulting in highest prio is a bit surprising.